### PR TITLE
refactor: create VDR method to be generic

### PR DESCRIPTION
**Pull request by fqutishat**
Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

### CODE REVIEW

File: pkg/framework/aries/api/vdr/vdr.go [hunk 1]
```diff
@@ -24,17 +23,96 @@ const DIDCommServiceType = "did-communication"
...
 }
 
 // VDR verifiable data registry interface.
 type VDR interface {
-	Read(did string, opts ...resolve.Option) (*did.DocResolution, error)
-	Store(doc *did.Doc, by *[]doc.ModifiedBy) error
-	Build(keyManager kms.KeyManager, opts ...create.Option) (*did.DocResolution, error)
+	Read(did string, opts ...ResolveOption) (*did.DocResolution, error)
+	Store(doc *did.Doc, by *[]ModifiedBy) error
+	Build(keyManager kms.KeyManager, did *did.Doc, opts ...CreateOption) (*did.DocResolution, error)
```
Comment (troyronda) [inline pkg/framework/aries/api/vdr/vdr.go:36]: We should have a TODO to rename this function.

Comment (troyronda) [inline pkg/framework/aries/api/vdr/vdr.go:36]: We should have a TODO to rename this function.

File: pkg/framework/aries/api/vdr/vdr.go [hunk 2]
```diff
@@ -24,17 +23,96 @@ const DIDCommServiceType = "did-communication"
...
 	Close() error
 }
 
 // VDR verifiable data registry interface.
 type VDR interface {
-	Read(did string, opts ...resolve.Option) (*did.DocResolution, error)
-	Store(doc *did.Doc, by *[]doc.ModifiedBy) error
-	Build(keyManager kms.KeyManager, opts ...create.Option) (*did.DocResolution, error)
+	Read(did string, opts ...ResolveOption) (*did.DocResolution, error)
+	Store(doc *did.Doc, by *[]ModifiedBy) error
```
Comment (troyronda) [inline pkg/framework/aries/api/vdr/vdr.go:35]: We should have a TODO to resolve the difference between Store and Create.

File: pkg/framework/aries/api/vdr/vdr.go (hunk 1 above)
Comment (fqutishat) [reply pkg/framework/aries/api/vdr/vdr.go:36]: Ok

Comment (fqutishat) [reply pkg/framework/aries/api/vdr/vdr.go:36]: Should I rename it to Create

File: pkg/framework/aries/api/vdr/vdr.go [hunk 3]
```diff
@@ -24,17 +23,96 @@ const DIDCommServiceType = "did-communication"
...
+
+// CreateOpts create did opts.
+type CreateOpts struct {
+	DefaultServiceType     string
+	DefaultServiceEndpoint string
+	EncryptionKey          *did.VerificationMethod
+}
+
+// CreateOption is a create DID option.
+type CreateOption func(opts *CreateOpts)
```
Comment (troyronda) [inline pkg/framework/aries/api/vdr/vdr.go:94]: How do these Create options intersect with the fact that a DID document is being passed-in?

File: pkg/framework/aries/api/vdr/vdr.go (hunk 1 above)
Comment (troyronda) [reply pkg/framework/aries/api/vdr/vdr.go:36]: yes - suggest matching the function names in did-core.

Comment (troyronda) [reply pkg/framework/aries/api/vdr/vdr.go:36]: I think the CreateOptions go with the `Build` function not the `Create` function.

Comment (fqutishat) [reply pkg/framework/aries/api/vdr/vdr.go]: I removed CreateOptions

File: pkg/framework/aries/api/vdr/vdr.go (hunk 2 above)
Comment (fqutishat) [reply pkg/framework/aries/api/vdr/vdr.go]: I rename to create

File: pkg/framework/aries/api/vdr/vdr.go [hunk 4]
```diff
@@ -24,17 +23,70 @@ const DIDCommServiceType = "did-communication"
...
+	Read(did string, opts ...ResolveOption) (*did.DocResolution, error)
+	Create(keyManager kms.KeyManager, did *did.Doc, opts ...DIDMethodOption) (*did.DocResolution, error)
 	Accept(method string) bool
 	Close() error
 }
+
+// ResolveOpts holds the options for did resolve.
+type ResolveOpts struct {
+	HTTPClient  *http.Client
+	VersionID   interface{}
```
Comment (sandrask) [inline pkg/framework/aries/api/vdr/vdr.go:43]: Please add todo to see if we can remove VersionID and VersionTime from here.

File: pkg/framework/aries/api/vdr/vdr.go (hunk 3 above)
Comment (fqutishat) [reply pkg/framework/aries/api/vdr/vdr.go]: I removed CreateOptions

File: pkg/framework/aries/api/vdr/vdr.go (hunk 4 above)
Comment (fqutishat) [reply pkg/framework/aries/api/vdr/vdr.go:42]: done

File: pkg/didcomm/protocol/didexchange/states.go [hunk 5]
```diff
@@ -564,8 +559,13 @@ func (ctx *context) resolveDidDocFromConnection(conn *Connection) (*did.Doc, err
...
 	}
 
+	didMethod, err := vdr.GetDidMethod(didDoc.ID)
+	if err != nil {
+		return nil, err
+	}
+
 	// store provided did document
-	err := ctx.vdRegistry.Store(didDoc)
+	_, err = ctx.vdRegistry.Create(didMethod, didDoc, vdrapi.WithOption("store", true))
```
Comment (sandrask) [inline pkg/didcomm/protocol/didexchange/states.go:568]: It seems redundant to call get did method first in order to call create?

File: pkg/vdr/peer/creator.go [hunk 6]
```diff
@@ -7,89 +7,118 @@ SPDX-License-Identifier: Apache-2.0
...
+
+		store, ok = storeOpt.(bool)
+		if !ok {
+			return nil, fmt.Errorf("store opt not boolean")
+		}
 	}
 
-	return docResolution, nil
+	if !store {
+		docResolution, err := build(keyManager, didDoc, docOpts)
```
Comment (sandrask) [inline pkg/vdr/peer/creator.go:45]: Can you build and store at the same time?

**Review by sandrask — APPROVED** [review]

**Review by troyronda — APPROVED** [review]

**Pull request merged**